### ISSUE TITLE
Combining Enums at Comptime Crashes Compiler

### ISSUE THREAD

**Issue by nsmryan** [ziglang/zig#13606]
### Zig Version

0.11.0-dev.4799+8c4faa5f3

### Steps to Reproduce and Observed Behavior

Running the following program with 'zig run' crashes the compiler:
```zig
const std = @import("std");

pub const CombinedEnum = blk: {
    const numFields = @typeInfo(FirstEnum).Enum.fields.len + @typeInfo(SecondEnum).Enum.fields.len;
    comptime var fields: [numFields]std.builtin.Type.EnumField = undefined;

    comptime var index = 0;
    for (@typeInfo(FirstEnum).Enum.fields) |field| {
        fields[index] = field;
        index += 1;
    }

    for (@typeInfo(SecondEnum).Enum.fields) |field| {
        fields[index] = field;
        index += 1;
    }

    const enumInfo = std.builtin.Type.Enum{
        .layout = std.builtin.Type.ContainerLayout.Auto,
        .tag_type = u8,
        .fields = &fields,
        .decls = &[0]std.builtin.Type.Declaration{},
        .is_exhaustive = true,
    };

    break :blk @Type(std.builtin.Type{ .Enum = enumInfo });
};

pub const FirstEnum = enum {
    field0,
};

pub const SecondEnum = enum {
    field1,
};

pub fn main() anyerror!void {
    std.debug.print("startup\n", .{});
    const first = CombinedEnum.field0;
    const second = CombinedEnum.field1;
    std.debug.print("first = {}, second = {}\n", .{ @enumToInt(first), @enumToInt(second) });
}
```

The error message is attached. 

[log.txt](https://github.com/ziglang/zig/files/10050184/log.txt)

The error log indicates that an assert in array_hash_map.zig in putAssumeCapacityNoClobberContext is getting hit.

On 0.10.0-dev.4707+209a0d2a8 this is a segfault, so I posted this with the Zig version that does at least print an error trace.

### Expected Behavior

The expected behavior is to create an enum combining the fields of FirstEnum and SecondEnum.
I would expect a printout like:
```
first = 0, second = 1
```

**Comment by Vexu** [comment]
#12805 for generated enums, fix can be copied from #12817

**Comment by nsmryan** [comment]
I want to point out that from reading those linked issues I realize that the CombinedEnum was using the same values for multiple variants, as it used the defaults for each of FirstEnum.field0 (which is 0), and SecondEnum.field1 (which is also 0).

I did not intend this, and manually setting distinct enum values works just fine. 
I copied in the corrected code below for reference in case someone else ends up here and doesn't realize the underlying mistake.
```
const std = @import("std");

pub const CombinedEnum = blk: {
    const numFields = @typeInfo(FirstEnum).Enum.fields.len + @typeInfo(SecondEnum).Enum.fields.len;
    comptime var fields: [numFields]std.builtin.Type.EnumField = undefined;

    comptime var index = 0;
    for (@typeInfo(FirstEnum).Enum.fields) |field| {
        fields[index] = std.builtin.Type.EnumField{ .name = field.name, .value = index };
        index += 1;
    }

    for (@typeInfo(SecondEnum).Enum.fields) |field| {
        fields[index] = std.builtin.Type.EnumField{ .name = field.name, .value = index };
        index += 1;
    }

    const enumInfo = std.builtin.Type.Enum{
        .layout = std.builtin.Type.ContainerLayout.Auto,
        .tag_type = u8,
        .fields = &fields,
        .decls = &[0]std.builtin.Type.Declaration{},
        .is_exhaustive = true,
    };

    break :blk @Type(std.builtin.Type{ .Enum = enumInfo });
};

pub const FirstEnum = enum {
    field0,
};

pub const SecondEnum = enum {
    field1,
};

pub fn main() anyerror!void {
    std.debug.print("startup\n", .{});
    const first = CombinedEnum.field0;
    const second = CombinedEnum.field1;
    std.debug.print("first = {}, second = {}\n", .{ @enumToInt(first), @enumToInt(second) });
}
```

Thank you Vexu.
